src/main/flight/autopilot.c

```diff
@@ -91,17 +215,186 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+            anglesBodyFrame = (vector2_t){{0, 0}};             // set output PIDS to 0; upsampling filter will smooth this
+            // reset target location each cycle (and set previousDistance to zero in for loop), to keep D current, and avoid a spike when stopping
+            ap.targetLocation = gpsSol.llh;
+            // keep updating sanity check distance while sticks are out because speed may get high
+            ap.sanityCheckDistance = sanityCheckDistance(gpsSol.groundSpeed);
+            ap.axesMoving = 2; // keep the counter at 2 while sticks are moving
+        } else {
+            // ** Rotate pid Sum to body frame, and convert it into pitch and roll **
+            // attitude.values.yaw increases clockwise from north
+            // PID is running in ENU, adapt angle (to 0deg = EAST);
+            //  rotation is from EarthFrame to BodyFrame, no change of sign from heading
+            const float angle = DECIDEGREES_TO_RADIANS(attitude.values.yaw - 900);
+            vector2_t pidBodyFrame;   // pid output in body frame; X is forward, Y is left
+            vector2Rotate(&pidBodyFrame, &pidSum, angle);         // rotate by angle counterclockwise
+            anglesBodyFrame.v[AI_ROLL] = -pidBodyFrame.y;         // negative roll to fly left
+            anglesBodyFrame.v[AI_PITCH] = pidBodyFrame.x;         // positive pitch for forward
+             // limit angle vector to maxAngle
+            const float mag = vector2Norm(&anglesBodyFrame);
+            if (mag > ap.maxAngle && mag > 0.0f) {
+                vector2Scale(&anglesBodyFrame, &anglesBodyFrame, ap.maxAngle / mag);
+            }
+        }
+        ap.pidSumBodyFrame = anglesBodyFrame;    // this value will be upsampled
+    }
+
+    // Final output to pid.c Angle Mode at 100Hz with PT3 upsampling
+    for (unsigned i = 0; i < RP_AXIS_COUNT; i++) {
+        // note: upsampling should really be done in earth frame, to avoid 10Hz wobbles if pilot yaws and the controller is applying significant pitch or roll
+        autopilotAngle[i] = pt3FilterApply(&ap.upsampleLpfBodyFrame[i], ap.pidSumBodyFrame.v[i]);
+    }
+
+    if (debugAxis < 2) {
+        // this is different from @ctzsnooze version
+        // debugAxis = 0: store longitude + roll
+        // debugAxis = 1: store latitude + pitch
+        DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 1, lrintf(debugGpsDistance.v[debugAxis]));    // cm
+        DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 2, lrintf(debugPidSumEF.v[debugAxis] * 10));  // deg * 10
+        DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 3, lrintf(autopilotAngle[debugAxis] * 10));   // deg * 10
+    }
+    return true;
+}
+
+bool isBelowLandingAltitude(void)
+{
+    return getAltitudeCm() < 100.0f * apConfig()->landing_altitude_m;
 }
 
 float getAutopilotThrottle(void)
 {
     return throttleOut;
 }
+
+bool isAutopilotActive(void)
+{
+    return !ap.sticksActive;
+}
```

Comment:
Just a note:

This looks "wrong" in the sense that it doesn't really check if the autopilot mode is engaged, rather if sticks are being moved or not. Even if sticks are being moved, the autopilot is active as long as some autonomous mode is currently active.

It just looks like this is some very specialized logic for a very special code part, so exposing this for use in all of BF seems odd (what if someone in the future actually wants to check if the autopilot is active and not if sticks are being moved?)